### DISCUSSION TITLE
Need some guidance on formatter configurator

### DISCUSSION THREAD

**Discussion by efast1568** [s9e/TextFormatter#237] [Support]
Hi Joshy,

Once again, I'm reaching out for some assistance.

ChatGPT helped me set up exclusion rules for the img class in the autolink feature, and it's working as expected.

Here’s the relevant code in the Configurator.php file:

```
/www/wwwroot/waterhole/forum/vendor/s9e/text-formatter/src/Plugins/Autolink/Configurator.php

protected function getRegexp()
{
    $anchor = RegexpBuilder::fromList($this->configurator->urlConfig->getAllowedSchemes()) . ':';
    if ($this->matchWww)
    {
        $anchor = '(?:' . $anchor . '|www\\.)';
    }

    $excludedFormats = '\.(?:jpg|jpeg|png|gif|bmp|svg|webp)';

    $regexp = '#\\b' . $anchor . '(?!.*' . $excludedFormats . ')(?>[^\\s()\\[\\]'
            . '\\x{FF01}-\\x{FF0F}\\x{FF1A}-\\x{FF20}\\x{FF3B}-\\x{FF40}\\x{FF5B}-\\x{FF65}'
            . ']|\\([^\\s()]*\\)|\\[\\w*\\])++#Siu';

    return $regexp;
}
```
I’d like to ask if it's possible to achieve the same effect by configuring it in the ServiceProvider.php file. Here is my current setup. I've tried making some adjustments for autolink, but so far, I've not been successful.

```
public function boot()
{
    Extend\Formatter::configure(function (Configurator $configurator) {
        // Adding settings for media embedding
        $configurator->MediaEmbed->add('facebook');
        $configurator->MediaEmbed->add('youtube');
        $configurator->MediaEmbed->add('instagram');
        $configurator->MediaEmbed->add('tiktok');
        $configurator->MediaEmbed->add('twitter');
        $configurator->MediaEmbed->add('imgur');
        $configurator->MediaEmbed->add(
            'bilibili',
            [
                'host' => ['bilibili.com', 'b23.tv',],
                'extract' => [
                    "!bilibili\.com/video(/av(?'aid'[-0-9]+))|(/BV(?'bvid'[-0-9A-Z_a-z]+))(\?p=(?'pn'[-0-9]+))?!",
                    "!b23\.tv/(/av(?'aid'[-0-9]+))|(/BV(?'bvid'[-0-9A-Z_a-z]+))(/p(?'pn'[-0-9]+))?!"
                ],
                'iframe' => [
                    'src' => '//player.bilibili.com/player.html?aid={@aid}&bvid={@bvid}&page={@pn}&autoPlay=0'
                ]
            ]
        );
        $configurator->MediaEmbed->add(
            'acfun',
            [
                'host' => 'acfun.cn',
                'extract' => "!acfun\.cn/v/ac(?'acid'[-0-9]+)!",
                'iframe' => [
                    'src' => '//www.acfun.cn/player/ac{@acid}'
                ]
            ]
        );
        $configurator->MediaEmbed->add(
            'ixigua',
            [
                'host' => 'ixigua.com',
                'extract' => "!ixigua\.com/(?'id'[-0-9]+)!",
                'iframe' => [
                    'src' => 'https://www.ixigua.com/iframe/{@id}?autoPlay=0',
                ]
            ]
        );

        // Redefining the IMG tag template to include FancyBox-related attributes
        $configurator->tags['IMG']->template = <<<xsl
            <img 
                data-fancybox="gallery" 
                src="{@src}" 
                alt="{@alt}" 
                title="{@title}" 
                width="{@width}" 
                height="{@height}" 
                style="cursor: default;">
            </img>
        xsl;
    });
}
```
Perhaps you could guide me in the right direction.

Much appreciated!

BR,
Kai

**Comment by JoshyPHP** [comment]
If the goal is to replace image URLs with the image itself, you can use the [Autoimage plugin](https://s9etextformatter.readthedocs.io/Plugins/Autoimage/Synopsis/). Otherwise, I'd probably create a tag filter that invalidates the tag if the URL matches a static image and that no markup was used to create the link, so that you can still manually/explicitly link to those.

https://s9etextformatter.readthedocs.io/Filters/Callback_signature/

```php
$configurator = new s9e\TextFormatter\Configurator;

function unlinkStaticImages(s9e\TextFormatter\Parser\Tag $tag, $tagText)
{
	if ($tagText === '' && preg_match('(\\.(?:bmp|gif|jpe?g|png|svg|webp)$)', $tag->getAttribute('url')))
	{
		$tag->invalidate();
	}
}

$configurator->Autolink;
$configurator->Litedown;
$configurator->tags['URL']->filterChain->append('unlinkStaticImages($tag, $tagText)')->setJS('
	function (tag, tagText)
	{
		if (tagText === "" && /\\.(?:bmp|gif|jpe?g|png|svg|webp)$/.test(tag.getAttribute("url"))
		{
			tag.invalidate();
		}
	}
');

extract($configurator->finalize());

$text = "http://localhost/img.png\n"
      . "http://localhost/img.php\n"
      . '<http://localhost/img.png>';
$xml  = $parser->parse($text);
$html = $renderer->render($xml);

die("$html\n");
```
```html
<p>http://localhost/img.png
<a href="http://localhost/img.php">http://localhost/img.php</a>
<a href="http://localhost/img.png">http://localhost/img.png</a></p>
```

**Reply by efast1568** [reply]
Tks Sir
 I will try that further. My issue is that when autoimage is enabled and autolink works together with fancybox, a simple image URL produces both '<a href= ' and '<img src 'tags, resulting in duplicate linking. To get just the <img> tag, you need to use Markdown syntax. That’s why I want to exclude image tags from autolink.

**Comment by JoshyPHP** [comment]
Not sure whether it's better, but you can use the `Preg` plugin to force portions of the text that match a regexp to be assign to a specific tag. For example, you can force anything that looks like a static image URL to match the `IMG` tag. That's basically what the `Autoimage` plugin does, except that tags created by the `Preg` plugin have precedence over those from other plugins. Text consumed by the `Preg` plugin won't be available for `Autoimage` or `Autolink`, invalidating their tags. To put it more simply, tags created by the `Preg` plugin override most other tags.

```php
$configurator->Preg->match(
	'#\\b(?<src>https?://[\\w./]+\\.(?:gif|png))(?!\\S)#i',
	'IMG'
);
```

1. https://s9etextformatter.readthedocs.io/Plugins/Preg/Synopsis/#using-match
2. https://s9etextformatter.readthedocs.io/Internals/Tag_priorities/

**Reply by efast1568** [reply]
As expected, only Joshy, the code deity, could have provided this.

This is exactly what I needed.

// Use the Preg plugin to match static image URLs that fit the regular expression and mark them as IMG tags


```
$configurator->Preg->match(
    '#(?<src>https?://[\\w./%-]+\\.(?:bmp|jpe?g|png|webp))#i',
    'IMG'
);
```

Now I understand—it was never the teacher's lack of ability, but rather the student's poorly phrased question.

Wishing you a pleasant day!
